### ISSUE TITLE
vktrace/replay swapchain handling is broken for DOTA2 Linux

### ISSUE THREAD

**Issue by eero-t** [LunarG/VulkanTools#154]
Setup:
- HW: SKL GT2
- OS: Ubuntu 16.04 with all the updates
- SW: Latest VulkanTools & Mesa from git

Use-case:
- Start vktrace in server mode: vktrace -o dota.vktrace
- Start DOTA2 with tracing layer and "-vulkan" option
- Start (Mechanics) tutorial once game loads

Expected result:
- Game works fine and rendering looks same as when tracing is not enabled

Actual result:
- User sees constantly wrong frames during the tutorial, ones that can be several seconds old (tracing slows the game down, so this means user being shown frame that is tens of frames from the past)

Notes:
- Replay of the trace always crashes early on DOTA2 startup due to invalid swapchain index.  For backtraces and some analysis, see bug #121 
- The invalid swapchain entry during replay is (always) 64th, which might correspond to to user seeing way too old frames during tracing.  Maybe it's negative offset wrapup?
- If one runs vkreplay with validation, there are lot of warnings about bogus semaphore operations which may also be related.  Not sure whether these are related to bug #152 
- By running vkreplay with validation or something else that slows it down, replay can run to the end.  This indicates that the issue is timing sensitive, e.g. race condition.  -> It's possible that on other platforms DOTA2 tracing and/or replay works only by luck
- Few weeks ago: Tuturial never started with tracing, loading animation just continued and trace grew infinitely

**Comment by eero-t** [comment]
Semaphore issue is open as item #100 (test-case is from VulkanSamples repo).

**Comment by eero-t** [comment]
With latest VulkanTools I was for the first time able to do a recording and replay without any issues with Mesa.  There are still about same validation warnings, but replay works both with and without validation.

The trace files were much smaller (I noticed commit improving trace content change detection) and replay was almost as fast as the real thing (1/3 slower instead of 3x slower).

Great progress, this works now much better than Apitrace for GL version of DOTA2!

**Comment by eero-t** [comment]
When I compiled vkreplay with -O3 instead of the default -O2, which increased the replay speed a bit, I got again crashes at replay start (9/10 of the runs).  This wasn't a compilation option related issue because I get a crash with default -O2 build also (maybe 1/5 of runs).
```
DS(ERROR): object: 0x1 type: 5 location: 4793 msgCode: 77: Queue 0x0x2176b80 is waiting on semaphore 0x1 that has no way to be signaled.
vkreplay error: Return value VK_ERROR_OUT_OF_DATE_KHR from API call (vkAcquireNextImageKHR) does not match return value from trace file VK_SUCCESS.
vkreplay error: Failed to replay packet_id 163, with global_packet_index 53231.
vkreplay error: Return value VK_ERROR_OUT_OF_DATE_KHR from API call (vkAcquireNextImageKHR) does not match return value from trace file VK_SUCCESS.
vkreplay error: Failed to replay packet_id 163, with global_packet_index 56441.
DS(ERROR): object: 0x1 type: 5 location: 4793 msgCode: 77: Queue 0x0x2176b80 is waiting on semaphore 0x1 that has no way to be signaled.
DS(ERROR): object: 0x220ad40 type: 27 location: 12116 msgCode: 90: vkQueuePresentKHR: Swapchain image index too large (140). There are only 3 images in this swapchain.
Segmentation fault (core dumped)
```

**Comment by eero-t** [comment]
When replaying ~33GB vktrace of DOTA2 finals gameplay recording (*), there's something funky in the DOTA2 startup phase.  That startup takes **~3 minutes** with vkreplay, with frames being output only at ~10s interval (+ GPU suspended >99% of time) and vkreplay being fully IO-bound (constantly in D-state according to "top").

With DOTA2 game itself, that startup takes only 1/2 minutes, and after few seconds, outputs frames every second.

The funky thing is that if vkreplay IO would be due to loading of game assets, I would assume vkreplay memory usage to grow, but it's not (according VmSize & Private Dirty totals reported by /proc/PID/smaps).

There are a lot of validation errors with vkreplay if core validation is enabled (without validation, it crashes at start 9/10 times). I see also several of these warnings when the teams screen comes up:
`vkreplay error: gpuMemory::copyMappingData() null src pointer.`

According to "perf", most of the active CPU cycles during the 3 mins is spent by kernel doing work requested by vkreplay:
```
# Overhead  Command Shared Object        Symbol        
    10.50%  vkreplay [kernel.kallsyms]    [k] copy_user_enhanced_fast_string       
    10.49%  vkreplay [unknown]            [k] 0x00007fb83b691566 
     7.91%  vkreplay libc-2.23.so         [.] __memcpy_avx_unaligned      
     3.74%  vkreplay [kernel.kallsyms]    [.] native_irq_return_iret      
     2.06%  swapper  [kernel.kallsyms]    [k] intel_idle
     1.90%  vkreplay [kernel.kallsyms]    [k] put_page  
     1.79%  vkreplay [kernel.kallsyms]    [k] copy_page_to_iter_iovec     
     1.30%  kswapd0  [kernel.kallsyms]    [k] shrink_page_list   
     1.23%  vkreplay [kernel.kallsyms]    [k] find_get_entry     
     1.12%  vkreplay [kernel.kallsyms]    [k] __wake_up_bit      
     1.11%  vkreplay [kernel.kallsyms]    [k] __radix_tree_lookup
     1.04%  vkreplay [kernel.kallsyms]    [k] unlock_page        
     0.99%  vkreplay [kernel.kallsyms]    [k] get_page_from_freelist      
     0.99%  vkreplay [kernel.kallsyms]    [k] i915_gem_object_get_dirty_page       
     0.86%  vkreplay [kernel.kallsyms]    [k] handle_mm_fault    
     0.75%  swapper  [kernel.kallsyms]    [k] __wake_up_bit      
     0.71%  vkreplay [kernel.kallsyms]    [k] ext4_mpage_readpages
```

(*) http://www.phoronix-test-suite.com/benchmark-files/dota2-pts-1971360796.dem.tar.bz2

**Comment by davidlunarg** [comment]
There is at least one know problem with vktrace and Dota2: sometimes mapped memory isn't correctly copied to the trace file. I have seen this on nVidia graphics, I don't know if it occurs with other graphics. This might be the cause of the copyMappingData error messages. I'm working on this problem.

Is it possible your replay is disk I/O bound? One way to determine that is to create a ram disk, and have vktrace create the file on the ram disk, and play it back from the ram disk. I saw a big performance difference when using the ram disc.

**Comment by eero-t** [comment]
Ramfs/tmpfs store their data in page cache, so it's kind of redundant when you can just redo the test-case. Using tmpfs for the recording didn't make replay go any faster, than just replaying the file multiple times, which is expected.  iostat says there's 0kB read from real disk during successive replays, when there's enough RAM for keeping the the trace file in page cache.

If trace is way too large to fit into RAM (or RAM disk), like above referred one, then sure, it can be partly disk bound on every run.  Above replay is occasionally in D state and 15-30% in IO-wait, but kernel doesn't tell how much of the wait is for disk or GPU.  I don't think that explains the issue.

**Comment by eero-t** [comment]
Tried again with latest Mesa, DOTA2 & VulkanTools.  I wasn't able to reproduce **vkreplay** swapchain issue with -O2 or -O3 compiled vkreplay, with or without validation, or running vkreplay under strace or valgrind.

However, I got the warning & crash every time when I test the same trace with **vktraceviewer**, when running that under validation:
```
ObjectTracker(ERROR): object: 0x3d75 type: 5 location: 374 msgCode: 146: Invalid Semaphore Object 0x3d75. For more information refer to Vulkan Spec Section '5.4. Command Buffer Submission' which states 'If waitSemaphoreCount is not 0, pWaitSemaphores must be a pointer to an array of waitSemaphoreCount valid VkSemaphore handles' (https://www.khronos.org/registry/vulkan/specs/1.0-extensions/xhtml/vkspec.html#VkSubmitInfo)
DS(ERROR): object: 0x3ee1 type: 5 location: 4545 msgCode: 78: Queue 0x0x7ffecc30c170 is waiting on semaphore 0x3ee1 that has no way to be signaled.
DS(ERROR): object: 0x3f80 type: 5 location: 4545 msgCode: 78: Queue 0x0x7ffecc30c170 is waiting on semaphore 0x3f80 that has no way to be signaled.
DS(ERROR): object: 0x3ffc type: 5 location: 4545 msgCode: 78: Queue 0x0x7ffecc30c170 is waiting on semaphore 0x3ffc that has no way to be signaled.
DS(ERROR): object: 0x41e9 type: 5 location: 4545 msgCode: 78: Queue 0x0x7ffecc30c170 is waiting on semaphore 0x41e9 that has no way to be signaled.
DS(ERROR): object: 0x426b type: 5 location: 4545 msgCode: 78: Queue 0x0x7ffecc30c170 is waiting on semaphore 0x426b that has no way to be signaled.
DS(ERROR): object: 0x43d8 type: 5 location: 4545 msgCode: 78: Queue 0x0x7ffecc30c170 is waiting on semaphore 0x43d8 that has no way to be signaled.
DS(ERROR): object: 0x4448 type: 5 location: 4545 msgCode: 78: Queue 0x0x7ffecc30c170 is waiting on semaphore 0x4448 that has no way to be signaled.
DS(ERROR): object: 0x44be type: 5 location: 4545 msgCode: 78: Queue 0x0x7ffecc30c170 is waiting on semaphore 0x44be that has no way to be signaled.
DS(ERROR): object: 0x4596 type: 5 location: 4545 msgCode: 78: Queue 0x0x7ffecc30c170 is waiting on semaphore 0x4596 that has no way to be signaled.
DS(ERROR): object: 0x5167 type: 5 location: 4545 msgCode: 78: Queue 0x0x7ffecc30c170 is waiting on semaphore 0x5167 that has no way to be signaled.
DS(ERROR): object: 0x86 type: 27 location: 10629 msgCode: 95: vkQueuePresentKHR: Swapchain image index too large (20839). There are only 3 images in this swapchain.

Thread 4 "ReplayThread" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffed3882700 (LWP 25811)]
0x00007ffff22a5c77 in xshmfence_reset () from /usr/lib/x86_64-linux-gnu/libxshmfence.so.1
```

Is there some difference in swapchain handling between vkreplay and vktraceviewer, or do they use exactly the same code?  Except for the swapchain issues, validation errors seem same as what I get under vkreplay, and vkreplay doesn't anymore crash.

Above happens with VK_LAYER_LUNARG_standard_validation, on start of first replay.  With VK_LAYER_LUNARG_core_validation I got different issue on second replay round:
```
DS(ERROR): object: 0x7ffecc321380 type: 27 location: 10640 msgCode: 96: vkQueuePresentKHR: Swapchain image index 1 has not been acquired.
DS(ERROR): object: 0x7ffeb4bfd3d0 type: 3 location: 10332 msgCode: 2334: vkCreateSwapChainKHR() called with imageExtent = (1920,1080), which is not equal to the currentExtent = (1855,1056) returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR(). For more information refer to Vulkan Spec Section '30.6. WSI Swapchain' which states 'imageExtent must be between minImageExtent and maxImageExtent, inclusive, where minImageExtent and maxImageExtent are members of the VkSurfaceCapabilitiesKHR structure returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR for the surface' (https://www.khronos.org/registry/vulkan/specs/1.0-extensions/xhtml/vkspec.html#VkSwapchainCreateInfoKHR)

Thread 4 "ReplayThread" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffed3882700 (LWP 25901)]
0x00007ffed26ab092 in core_validation::CreateFramebuffer(VkDevice_T*, VkFramebufferCreateInfo const*, VkAllocationCallbacks const*, VkFramebuffer_T**) () from build/layers/./libVkLayer_core_validation.so
(gdb) bt
#0  0x00007ffed26ab092 in core_validation::CreateFramebuffer(VkDevice_T*, VkFramebufferCreateInfo const*, VkAllocationCallbacks const*, VkFramebuffer_T**) () from build/layers/./libVkLayer_core_validation.so
#1  0x0000000000449c4e in vkReplay::manually_replay_vkCreateFramebuffer(packet_vkCreateFramebuffer*) ()
#2  0x00000000004582d8 in vkReplay::replay(vktrace_trace_packet_header*) ()
#3  0x0000000000442e88 in VkReplayReplay(vktrace_trace_packet_header*) ()
#4  0x00000000004325bd in vktraceviewer_QReplayWorker::playCurrentTraceFile(unsigned long) ()
```

Topping the trace info window instead of the trace replay window seems to be something that triggers crashes.

Btw. Vktraceviewer itself shows the this error, but it happens on first replay:
`Error: Return value VK_ERROR_OUT_OF_DATE_KHR from API call (vkAcquireNextImageKHR) does not match value from trace file VK_SUCCESS.`

PS. Perf gap for the actual game play part to real DOTA2 was 15-25% even with -O3 version of vkreplay. Vkreplay is single threaded whereas DOTA2 (based on vktraveviewer output) does draws from 5 different threads.

**Comment by davidlunarg** [comment]
> Is there some difference in swapchain handling between vkreplay and vktraceviewer, or do they use exactly the same code? 

Some code is shared between vktraceviewer and vkreplay, but swapchain code looks like it isn't. It is likely that the swapchain code in the viewer was ported from replay, but they are not the same anymore.

> Vkreplay is single threaded whereas DOTA2 (based on vktraveviewer output) does draws from 5 different threads.

Yes, vkreplay is single threaded. We've thought about playing back using the same number of threads as the original app, but haven't spent the time yet on how it would be implemented.

**Comment by eero-t** [comment]
With latest DOTA2, Mesa and VulkanTools, DOTA2 replay fails now after lots of these messages:
`DS(ERROR): object: 0x1ded type: 23 location: 3016 msgCode: 61: Descriptor set 0x1ded encountered the following validation error at vkCmdDraw()() time: Descriptor in binding #0 at global descriptor index 0 is being used in draw but has not been updated.`

To this error:
`kreplay error: API call vkQueueSubmit returned VK_ERROR_DEVICE_LOST. vkreplay cannot continue, exiting.`

Before it showed any frame content on screen, there was just a black window (tried few times, same result every time).

-> I.e. cannot say whether the swapchain issue is still there.

**Comment by eero-t** [comment]
With latest DOTA2, Mesa and VulkanTools, DOTA2 replay freezes with 100% CPU utilization before showing any frames i.e. I cannot say whether the swapchain issues is still there.

Before that, API validation shows following errors:
```
DS(ERROR): object: 0xbaa5c0 type: 3 location: 10303 msgCode: 2334: vkCreateSwapChainKHR() called with imageExtent = (1920,1080), which is not equal to the currentExtent = (1855,1056) returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR(). For more information refer to Vulkan Spec Section '30.8. WSI Swapchain' which states 'imageExtent must be between minImageExtent and maxImageExtent, inclusive, where minImageExtent and maxImageExtent are members of the VkSurfaceCapabilitiesKHR structure returned by vkGetPhysicalDeviceSurfaceCapabilitiesKHR for the surface' (https://www.khronos.org/registry/vulkan/specs/1.0-extensions/html/vkspec.html#VkSwapchainCreateFlagBitsKHR)
../../../src/intel/isl/isl.c:1181: FINISHME: validate row pitch of stencil surfaces
IMAGE(ERROR): object: 0x0 type: 0 location: 2510 msgCode: 2172: vkCreateImageView() format VK_FORMAT_B8G8R8A8_UNORM differs from image 4 format VK_FORMAT_B8G8R8A8_SRGB.  Formats MUST be IDENTICAL unless VK_IMAGE_CREATE_MUTABLE_FORMAT BIT was set on image creation. For more information refer to Vulkan Spec Section '11.5. Image Views' which states 'If image was not created with the VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT flag, format must be identical to the format used to create image' (https://www.khronos.org/registry/vulkan/specs/1.0-extensions/html/vkspec.html#VkImageViewCreateInfo)
IMAGE(ERROR): object: 0x0 type: 0 location: 2510 msgCode: 2172: vkCreateImageView() format VK_FORMAT_B8G8R8A8_UNORM differs from image 5 format VK_FORMAT_B8G8R8A8_SRGB.  Formats MUST be IDENTICAL unless VK_IMAGE_CREATE_MUTABLE_FORMAT BIT was set on image creation. For more information refer to Vulkan Spec Section '11.5. Image Views' which states 'If image was not created with the VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT flag, format must be identical to the format used to create image' (https://www.khronos.org/registry/vulkan/specs/1.0-extensions/html/vkspec.html#VkImageViewCreateInfo)
IMAGE(ERROR): object: 0x0 type: 0 location: 2510 msgCode: 2172: vkCreateImageView() format VK_FORMAT_B8G8R8A8_UNORM differs from image 6 format VK_FORMAT_B8G8R8A8_SRGB.  Formats MUST be IDENTICAL unless VK_IMAGE_CREATE_MUTABLE_FORMAT BIT was set on image creation. For more information refer to Vulkan Spec Section '11.5. Image Views' which states 'If image was not created with the VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT flag, format must be identical to the format used to create image' (https://www.khronos.org/registry/vulkan/specs/1.0-extensions/html/vkspec.html#VkImageViewCreateInfo)
DS(ERROR): object: 0x1951 type: 23 location: 2799 msgCode: 61: Descriptor Set 0x1951 bound but was never updated. It is now being used to draw so this will result in undefined behavior.
DS(ERROR): object: 0x1951 type: 23 location: 2814 msgCode: 61: Descriptor set 0x1951 encountered the following validation error at vkCmdDraw() time: Descriptor in binding #0 at global descriptor index 0 is being used in draw but has not been updated.
```

**Comment by eero-t** [comment]
Closing, DOTA2 tracing seems to work fine now (I noticed there have been some fixes for swapchain handling in the meanwhile too).  If there are further issues later on, I'll open another issue.
